staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -279,16 +281,55 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		crdName = crdName + "core"
 	}
 
-	crdKey := crdName
-	clusterName, err := genericapirequest.ClusterNameFrom(ctx)
+	var crdClusterName string
+	var crd *apiextensionsv1.CustomResourceDefinition
+	var err error
+
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil {
 		responsewriters.ErrorNegotiated(
 			apierrors.NewInternalError(fmt.Errorf("error resolving resource: %v", err)),
 			Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, w, req,
 		)
+		return
+	}
+
+	if cluster.Wildcard {
+		// HACK: Search for the right logical cluster hosting the given CRD when watching or listing with wildcards.
+		// This is a temporary fix for issue https://github.com/kcp-dev/kcp/issues/183: One cannot watch with wildcards
+		// (across logical clusters) if the CRD of the related API Resource hasn't been added in the admin logical cluster first.
+		// The fix in this HACK is limited since the request will fail if 2 logical clusters contain CRDs for the same GVK
+		// with non-equal specs (especially non-equal schemas).
+		var crds []*apiextensionsv1.CustomResourceDefinition
+		crds, err = r.crdLister.List(labels.Everything())
+		if err == nil {
+			if len(crds) == 0 {
+				err = errors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
```

Comment:
would expect an empty set rather than a NotFound.

Reply:
Well, it seems to me that returning an empty set would mainly mean: this API exists but there is not resource of this API.
The situation here is that the API itself is unknown (no CRD is available), which afaict is reflected by a NotFound error.
But I realize that I should also trigger a NotFound error when no CRD was found after iterating the CRD list.

Reply:
sounds reasonable. The proxy will ignore NotFound if there is another shard that has the CRD.